app/assets/javascripts/components/InputWithIcon.tsx

```diff
@@ -0,0 +1,90 @@
+import { FunctionComponent, Ref } from 'preact';
+import { JSXInternal } from 'preact/src/jsx';
+import { forwardRef } from 'preact/compat';
+import { useState } from 'preact/hooks';
+import { Icon, IconType } from './Icon';
+import { IconButton } from './IconButton';
+
+type Props = {
+  icon: IconType;
+  inputType: 'text' | 'email' | 'password';
+  className?: string;
+  iconClassName?: string;
+  value: string;
+  onChange: JSXInternal.GenericEventHandler<HTMLInputElement>;
+  onFocus?: JSXInternal.GenericEventHandler<HTMLInputElement>;
+  onKeyDown?: JSXInternal.KeyboardEventHandler<HTMLInputElement>;
+  disabled?: boolean;
+  placeholder: string;
+  toggle?: {
+    toggleOnIcon: IconType;
+    toggleOffIcon: IconType;
+    title: string;
+    onClick: (toggled: boolean) => void;
+  };
+};
+
+const DISABLED_CLASSNAME = 'bg-grey-5 cursor-not-allowed';
+
+export const InputWithIcon: FunctionComponent<Props> = forwardRef(
+  (
+    {
+      icon,
+      inputType,
+      className,
+      iconClassName,
+      value,
+      onChange,
+      onFocus,
+      onKeyDown,
+      disabled,
+      toggle,
+      placeholder,
+    },
+    ref: Ref<HTMLInputElement>
+  ) => {
+    const [toggled, setToggled] = useState(false);
```

Comment:
Could pass `toggled` as part of the `toggle` prop to make this a fully controlled component

Reply:
As in, inside of having the `toggled` state inside the component, have it inside the parent?

Reply:
Wouldn't this cause a lot of repetition, since each component using this will have to have its own toggled state?

Reply:
Makes sense 👍️ I'll make the change.